[WIP] Property-based testing for snabb config

takikawa opened the pull request:
Initial property-based testing for snabb config, based on code by @kbara.

There are some unimplemented additions that would be useful such as generating `snabb config add` and `snabb config remove` commands. Also, the random value generation doesn't support all YANG types so it should probably be expanded (it supports enough for the snabb/ietf softwire protocols right now).

There are also some minor issues that I think can be fixed later, such as better value generation for list values.

Commit:
Property-based test: lwaftr run does not crash with get

This adds a first property based test. The property is that the lwaftr
does not crash with a 'snabb config get' command.

takikawa committed:
Minor fix to selftest.sh

Commit:
Close popens

Commit:
Add $ to exit code

Commit:
Run property-check lwaftr longer

Commit:
Sleep for longer before starting test

takikawa committed:
Generate xpath queries for `snabb config get` test

takikawa committed:
Add test ensuring two config runs yield the same result

takikawa committed:
Add cleanup code to shutdown lwaftr process for test

takikawa committed:
Produce "/" path instead of "" in base case

takikawa committed:
Abstract out the common prop testing code

takikawa committed:
Add generation of `snabb config set` commands

Use in the sameval property test

takikawa committed:
Add a new test for the get-state command

takikawa committed:
Make get-state lookups use generated queries too

takikawa committed:
Fix the get-state test to have non-empty query results

The empty environment for running the lwaftr was causing
the state query to return an empty result. The PATH variable
needs to be preserved at least.

Also adjust the xpath generation to accommodate the get-state
command.

takikawa committed:
Bump get-state test duration

takikawa committed:
Add optional query parameter to generate functions

Also change the API to export generation functions that
hardcode the right schema for these tests.

Use the optional parameter to fix the sameval property
(it wasn't actually setting the same property correctly)

takikawa committed:
Improve sameval test to be resilient to missing fields

Sometimes the generated queries are for fields that are
missing in the configuration. Ignore those cases for doing
a subsequent "set".

Also abstracted the crash handling code.

takikawa committed:
Abstract out the crash checking into common.lua

takikawa committed:
Generate selectors for list indexing in tests

takikawa committed:
Test the ietf-softwire config protocol too

takikawa committed:
Traverse 'list' nodes for random generation better

Also generate values for ipv6 addresses/prefixes and
unions to better explore the schema

takikawa committed:
Adjust how integer values are generated to prefer boundaries

takikawa committed:
Prefer lower position arguments

takikawa committed:
Start to add functions for generating set/add values

Generates values from the schema fo a given path for use with
the config set and add functions.

takikawa committed:
Use new set command generation for nocrash test

And bump test duration

takikawa committed:
Initial attempt at generating compound values

takikawa committed:
Generate more leaf data types

takikawa committed:
Add some comments to genyang.lua

takikawa committed:
Add comments, minor refactoring

kbara commented:
It's looking nice. Some bits, like the crash checks, look fragile but like a reasonably workable approach.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,369 @@
+end
+
+-- return a random number, preferring boundary values
+local function choose_range(lo, hi)
```
Preferring boundary values is good, but a weighting of 0.5 is probably a bit heavy - how would you feel about 0.1 or 0.2? I'm also inclined to include 0 in the list of numbers to test often.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,369 @@
+      return math.random(0, 255) .. "." .. math.random(0, 255) .. "." ..
+             math.random(0, 255) .. "." .. math.random(0, 255)
+   elseif prim == "ipv6-address" then
+      local addr = random_hexes()
```
This is a fine approach since it's a short string; for longer ones, it makes more sense to build up an array and then use a function that does the whole join in one go.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,369 @@
+      return str
+   end
+
+   -- TODO: generate these:
```
Looking forward to seeing those. :-)

kbara commented on src/program/lwaftr/tests/propbased/prop_sameval.lua:
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env luajit
```
Nice property. WDYT about occasionally hammering it with way more than 2 gets/sets, like 100-1000, maybe 1 time in 100?

kbara commented:
The current version LGTM, though I've flagged a couple of nits.

takikawa committed:
Adjust weight for boundary values in number ranges

takikawa committed:
Support decimal64 and binary types for test gen

takikawa committed:
Run more sameval get/set tests on occasion

takikawa committed:
Implement enum generation, change error behavior

takikawa committed:
Use range statements in YANG types for value gen

takikawa committed:
Validate PCI address for property based tests

takikawa committed:
Don't fail if there isn't a range in the type

takikawa committed:
Add code to occasionally ignore bounds

takikawa committed:
Intentionally generate wrong types occasionally

takikawa committed:
Experiment with `snabb config add` testing

Likely needs tweaking to be useful. The add command
only works on arrays/tables, so it should work harder
to generate the right paths for that.

takikawa committed:
Generate `snabb config remove` commands too

Similar to add, this needs to be more clever about
picking array/table nodes

takikawa committed:
Fix selftest for genyang

Don't generate values with invalid ranges or types
for the selftest to ensure the test is predictable.

takikawa committed:
Generate values/paths for config add/remove better

This commit fixes both the query paths and generated
values (for add) to test the add and remove commands
more thoroughly.

Also improves testing for the set command too.

takikawa committed:
Adjust leaf-list case to not always use a selector

takikawa commented:
Note: I've pushed the current state of this branch to the PR. There's one known issue I wanted to fix but didn't quite finish yet, which is that the result of `math.random` for some of the integer ranges (e.g., uint64) produces non-integers. So sometimes the test generates bad config fragments like `binding-table-version: 1.3115911891594e+16;`.

I plan to fix this by writing a random function that has the same interface as the two-argument `math.random` but divides the range in half and randomly selects a half to sample from (since `math.random` is fine on 32-bit ranges).

dpino commented:
@takikawa So if my understanding is correct the problem is that `math.randon(0, 2^64-1)` sometimes returns a number in float representation (makes sense since Lua numbers are 64-bit doubles), like 1.3115911891594e+16. As `binding-table-version` expects an uint64, would be incorrect to generate a configuration file with the value represented as a floating point value. Would it work to cast that value to an integer using an expression like?

```lua
> = string.format("%0.f", 1.3115911891594e+16)
13115911891594000
```

The result is an string, but if it's for generating a text configuration file, that would work I think.

kbara commented:
We can't generate the full range of 64 bit ints using a single call to math.random(). As Diego said, we're using LuaJIT with the native backing type being doubles. Your suggested workaround with combining two calls sounds fine to me, as long as you take care with handling the result, making sure it's returning a uint64 rather than a native Lua number and that no intermediate steps using Lua numbers require more precision than a native Lua number can provide.

kbara commented on src/program/lwaftr/tests/propbased/common.lua:
```diff
@@ -0,0 +1,52 @@
+     end
+  
+     -- TODO: validate the address
+     local pci_addr = prop_args[1]
```
Does this code already sufficiently address validation? If so, the TODO comment should be removed, I'd say.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,545 @@
+end
+
+-- choose a natural number (e.g., index or length of array) by
+-- repeating a cointoss
```
Why do you want to make it exponentially less likely to choose later options?

takikawa replied:
This is used often for generating the `n` in the `[position=n]` part of a `leaf-list` query. Since most of the lists used in the config are small, it's more useful to test smaller numbers to access data that's actually in the config.

Though it's also used for, e.g., picking the length of a random string. So another option would be to just bias, say, 1-5 and then pick a random number from a large range. I could implement that change if it seems better.

kbara replied:
Naw, sounds good as-is, thanks for the clarification.

kbara replied:
Or specifically, it sounds good as-is with the goal of merging it tomorrow; in the future, I'd like to see a way to generate longer strings with non-trivial probability.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,545 @@
+   elseif prim == "uint32" then
+      return choose_range(rng, 0, 4294967295)
+   elseif prim == "uint64" then
+      return choose_range(rng, 0, 18446744073709551615)
```
Wrong numeric type, use a LL suffix.

LuaJIT 2.1.0-beta1 -- Copyright (C) 2005-2015 Mike Pall. http://luajit.org/
JIT: ON SSE2 SSE3 SSE4.1 fold cse dce fwd dse narrow loop abc sink fuse
> print(18446744073709551615 == 18446744073709551614)
true
> print(18446744073709551615LL == 18446744073709551614LL)
false

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,545 @@
+   elseif prim == "int32" then
+      return choose_range(rng, -2147483648, 2147483647)
+   elseif prim == "int64" then
+      return choose_range(rng, -9223372036854775809, 9223372036854775807)
```
Use LL.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -0,0 +1,545 @@
+      return choose_range(rng, 0, 18446744073709551615)
+   -- TODO: account for fraction-digits and range
+   elseif prim == "decimal64" then
+      local int64 = math.random(-9223372036854775809, 9223372036854775807)
```
Use LL.

kbara commented on src/program/lwaftr/tests/propbased/selftest.sh:
```diff
@@ -0,0 +1,11 @@
+
+if [ -z $SNABB_PCI0 ]; then exit $SKIPPED_CODE; fi
+
+./snabb quickcheck program.lwaftr.tests.propbased.prop_nocrash $SNABB_PCI0
```
Ok for now, though ideally I'd like to be able to run these tests without a real NIC at some point, so I can just have my personal build server hammer away at it. Not a blocker for merging.

kbara commented:
What's the root cause of the test error?

```
src/testlog/program.lwaftr.tests.propbased.selftest.sh:
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 19790895
Warning: No NUMA memory affinity.
Pass --cpu to bind to a CPU and its NUMA node.
hugetlb mmap failed (Cannot allocate memory), falling back.
lib/ctable.lua:152: mmap failed: Cannot allocate memory
stack traceback:
	core/main.lua:149: in function <core/main.lua:147>
	[C]: in function 'error'
	lib/ctable.lua:152: in function 'calloc'
	lib/ctable.lua:167: in function 'resize'
	lib/ctable.lua:129: in function 'new'
	apps/lwaftr/ctable_wrapper.lua:41: in function 'initialize_frag_table'
	apps/lwaftr/ipv4_apps.lua:42: in function 'new'
	core/app.lua:305: in function <core/app.lua:304>
	core/app.lua:332: in function 'apply_config_actions'
	apps/config/follower.lua:50: in function 'commit_pending_actions'
	apps/config/follower.lua:62: in function 'handle_actions_from_leader'
	...
	program/lwaftr/setup.lua:507: in function 'run_worker'
	[string "require('program.lwaftr.setup').run_worker({[..."]:1: in function 'f'
	core/main.lua:58: in function <core/main.lua:43>
	[C]: in function 'xpcall'
	core/main.lua:211: in main chunk
	[C]: at 0x00452f60
	[C]: in function 'pcall'
	core/startup.lua:3: in main chunk
	[C]: in function 'require'
	[string "require "core.startup""]:1: in main chunk
100 iterations succeeded.
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 23502252
100 iterations succeeded.
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 26678293
link report:
Running the same config command twice produced different outputs
Git revision 41a3bcb6f3a64c50d7ceb7aeb1c8554eb4566c67
The property was falsified.
The command was:	./snabb config get -s ietf-softwire 6886 "/softwire-config/description"
Rerun as: snabb quickcheck --seed=26678293 --iterations=11 program.lwaftr.tests.propbased.prop_sameval 0000:82:00.0
EXITCODE: 1
```

kbara commented:
I currently see two blockers to merging:
a) The numeric issues - those should be straightforward to sort out
b) The test failure - the property based test suite always has the awkward compromise of needing to help us find bugs, but the version that runs in CI should not be expected to fail (ie, it should not trigger known bugs). Making our test suite less deterministic is generally a problem, and intermittent failures are best avoided. (That said, turning up a genuinely new bug at random during CI is fine!)

takikawa commented:
@kbara re: the error, you mean the quickcheck failure at the bottom rather than the hugetlb error right? It looks like that's after triggering one of those bursts of 100 gets at a time that you suggested (in the previous round of comments), and then causing the lwaftr or something else to get stuck during that. Looking into it.

takikawa committed:
Tweak test iterations and timeout

Having a large range of possible iterations makes the test
less predictable, so reduced the range.

takikawa committed:
Generate 64-bit ranges correctly

The 64-bit cases generated in this commit ignore the
`range` statements, however. In the current schemas,
no uint64/int64 types are used with range statements.

kbara commented on src/program/lwaftr/tests/propbased/genyang.lua:
```diff
@@ -123,6 +124,19 @@ local function random_hexes()
+   local r1 = ffi.cast("uint64_t", math.random(0, 2 ^ 32 - 1))
+   local r2 = ffi.cast("uint64_t", math.random(0, 2 ^ 32 - 1))
+
+   for i=1, 32 do
```
Why do you loop this instead of just multiplying by 4294967296LL?

takikawa replied:
Just a brain fart. Pushed a change for that.

takikawa committed:
Simplify the function in f4dae60e08c521c8

kbara commented:
LGTM

takikawa committed:
Delete outdated TODO comment
